networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/gossip/GossipHandler.java

```diff
@@ -13,112 +13,80 @@
   private static final Logger LOG = LogManager.getLogger();
+
+  private static CompletableFuture<Boolean> VALIDATION_FAILED =
+      CompletableFuture.completedFuture(false);
+  private static CompletableFuture<Boolean> VALIDATION_SUCCEEDED =
+      CompletableFuture.completedFuture(true);
   static final int GOSSIP_MAX_SIZE = 1048576;
   private static final int MAX_SENT_MESSAGES = 2048;
 
+  private final Topic topic;
   private final PubsubPublisherApi publisher;
-  private final EventBus eventBus;
-
+  private final TopicHandler handler;
   private final Set<Bytes> processedMessages =
       LimitedSet.create(MAX_SENT_MESSAGES, Mode.DROP_LEAST_RECENTLY_ACCESSED);
 
-  protected GossipTopicHandler(final PubsubPublisherApi publisher, final EventBus eventBus) {
+  public GossipHandler(
+      final Topic topic, final PubsubPublisherApi publisher, final TopicHandler handler) {
+    this.topic = topic;
     this.publisher = publisher;
-    this.eventBus = eventBus;
+    this.handler = handler;
   }
 
-  public abstract Topic getTopic();
-
   @Override
-  public final void accept(MessageApi message) {
+  public CompletableFuture<Boolean> apply(final MessageApi message) {
     Bytes bytes = Bytes.wrapByteBuf(message.getData()).copy();
     if (bytes.size() > GOSSIP_MAX_SIZE) {
       LOG.trace(
           "Rejecting gossip message of length {} which exceeds maximum size of {}",
           bytes.size(),
           GOSSIP_MAX_SIZE);
-      return;
+      return VALIDATION_FAILED;
     }
     if (!processedMessages.add(bytes)) {
       // We've already seen this message, skip processing
-      LOG.trace("Ignoring duplicate message for topic {}: {} bytes", getTopic(), bytes.size());
-      return;
+      LOG.trace("Ignoring duplicate message for topic {}: {} bytes", topic, bytes.size());
+      return VALIDATION_FAILED;
     }
-    LOG.trace("Received message for topic {}: {} bytes", getTopic(), bytes.size());
+    LOG.trace("Received message for topic {}: {} bytes", topic, bytes.size());
 
-    T data;
-    try {
-      data = deserializeData(bytes);
-      if (!validateData(data)) {
-        LOG.trace("Received invalid message for topic: {}", getTopic());
-        return;
-      }
-    } catch (SSZException e) {
-      LOG.trace("Received malformed gossip message on {}", getTopic());
-      return;
-    } catch (Throwable e) {
-      LOG.warn("Encountered exception while processing message for topic {}", getTopic(), e);
-      return;
-    }
-
-    // Post and re-gossip data on successful processing
-    gossip(bytes);
-    eventBus.post(data);
-  }
-
-  private T deserializeData(Bytes bytes) throws SSZException {
-    final T deserialized = deserialize(bytes);
-    if (deserialized == null) {
-      throw new SSZException("Unable to deserialize message for topic " + getTopic());
-    }
-    return deserialized;
+    final boolean result = handler.handleMessage(bytes);
+    return result ? VALIDATION_SUCCEEDED : VALIDATION_FAILED;
```

Comment:
Given this will be called on a Netty thread and we'll get a lot of gossip, should we be taking advantage of the fact that we return a CompletableFuture and do the validation on a separate thread?

Reply:
I think that makes sense.  Made a ticket to follow-up: https://pegasys1.atlassian.net/browse/BC-170